fop/fom.c

```diff
@@ -531,8 +534,12 @@ static void thr_addb2_leave(struct m0_loc_thread *thr,
 
 M0_INTERNAL void m0_fom_wakeup(struct m0_fom *fom)
 {
-	fom->fo_cb.fc_ast.sa_cb = readyit;
-	m0_sm_ast_post(&fom->fo_loc->fl_group, &fom->fo_cb.fc_ast);
+	struct m0_sm_ast *ast;
+	ast = m0_alloc(sizeof *ast);
```

Comment:
Can you elaborate on why this allocation is needed, why can't we use ast of fom itself ?

Reply:
```
709 /**
710  * Posts an AST to a group.
711  *
712  * An AST must not be re-posted until its previous (already posted) execution
713  * completes.
714  */
715 M0_INTERNAL void m0_sm_ast_post(struct m0_sm_group *grp, struct m0_sm_ast *ast);

```

This AST should not be re-posted or re-used again until its previous (already posted) execution completes.


But, here, the m0_fom_wakeup() may be called repeatedly and concurrently.
This is why the panic() in https://jts.seagate.com/browse/EOS-27717 happens.
This race condition is very rare, and only happens under heavy load.

If an AST has the potential to be re-used to post another callback, it has the possibility of panic, unless we make sure the AST is never re-used/re-posted before it is complete.

Reply:
More explain:
```
 135 M0_INTERNAL void m0_sm_ast_post(struct m0_sm_group *grp, struct m0_sm_ast *ast)
 136 {
 137         M0_PRE(ast->sa_cb != NULL);
 138         M0_PRE(ast->sa_next == NULL);
 139
 140         do {
 141                 ast->sa_next = grp->s_forkq;
 142                 /*
 143                 M0_LOG(M0_DEBUG, "grp=%p forkq_push: %p -> %p",
 144                        grp, ast, ast->sa_next);
 145                 */
 146         } while (!M0_ATOMIC64_CAS(&grp->s_forkq, ast->sa_next, ast));
 147         m0_clink_signal(&grp->s_clink);
 148 }
 149
 150 M0_INTERNAL void m0_sm_asts_run(struct m0_sm_group *grp)
 151 {
 152         struct m0_sm_ast *ast;
 153
 154         M0_PRE(grp_is_locked(grp));
 155
 156         while (1) {
 157                 do {
 158                         ast = grp->s_forkq;
 159                         /*
 160                         M0_LOG(M0_DEBUG, "grp=%p forkq_pop: %p <- %p",
 161                                grp, ast, ast->sa_next);
 162                         */
 163                 } while (ast != &eoq &&
 164                        !M0_ATOMIC64_CAS(&grp->s_forkq, ast, ast->sa_next));
 165
 166                 if (ast == &eoq)
 167                         break;
 168                 M0_ASSERT(ast->sa_next != NULL);
 169
 170                 ast->sa_next = NULL;
=====================context switch happens here================

In another thread, 
If we check the ast->sa_next, it can be NULL. 
If this AST is re-used in that thread, it will cause problems when context switch back...

================context switch back here=================
This thread will use `ast` which has already been modified.
==================================================

 171                 M0_ASSERT(!m0_is_poisoned(ast->sa_cb));
 172                 if (grp->s_addb2 == NULL) {
 173                         ast->sa_cb(grp, ast);
 174                 } else {
 175                         M0_ADDB2_HIST(grp->s_addb2->ga_forq,
 176                                       &grp->s_addb2->ga_forq_hist,
 177                                       m0_ptr_wrap(ast->sa_cb),
 178                                       ast->sa_cb(grp, ast));
 179                 }
 180         }
 181 }

```

Reply:
If the previous AST was posted for the same purpose, do we still need to post the new one ?

Reply:
If a new AST is allocated for every wakeup, we don't know whether the previous wakeup is already complete or not.

There is no harm to post the AST again. The callback function (wakeup_iff_waiting) will check if the FOM is already running.